internal/secretsource/url/url.go

```diff
@@ -0,0 +1,256 @@
+	// If empty, the entire response body is treated as the secret
+	ResponsePath string `json:"responsePath"`
+
+	// RequestsPerMinuteLimit sets the maximum requests per minute
+	RequestsPerMinuteLimit *int `json:"requestsPerMinuteLimit"`
+
+	// RequestsBurst allows a burst of requests above the rate limit
+	RequestsBurst *int `json:"requestsBurst"`
+
+	// Timeout for HTTP requests in seconds (defaults to 30)
+	TimeoutSeconds *int `json:"timeoutSeconds"`
+}
+
+const (
+	// The rate limiter replenishes tokens in the bucket once every 200 ms
+	// (5 per second) and allows a burst of 10 requests firing faster than
+	// that. If the client keeps making requests at the rapid pace, they
+	// will be slowed down. This allows a client to ask for a bunch of
+	// secrets at the start of a script, and then it slows it down to a
+	// reasonable pace.
+	defaultRequestsPerMinuteLimit = 300 // 300 requests per minute is one request every 200 ms
+	defaultRequestsBurst          = 10  // Allow a burst of 10 requests
+	defaultTimeoutSeconds         = 30  // 30 seconds timeout
+)
+
+//nolint:gochecknoinits // This is how k6 secret source registration works.
+func init() {
+	secretsource.RegisterExtension("url", func(params secretsource.Params) (secretsource.Source, error) {
+		config, err := getConfig(params.ConfigArgument, params.FS)
+		if err != nil {
+			return nil, fmt.Errorf("missing or invalid config: %w", err)
+		}
+
+		timeout := time.Duration(*config.TimeoutSeconds) * time.Second
+
+		return &urlSecrets{
+			config: config,
+			httpClient: &http.Client{
+				Timeout: timeout,
+			},
+			limiter: newLimiter(*config.RequestsPerMinuteLimit, *config.RequestsBurst),
+		}, nil
+	})
+}
+
+type urlSecrets struct {
+	config     extConfig
+	httpClient *http.Client
+	limiter    limiter
+}
+
+func (us *urlSecrets) Description() string {
+	return fmt.Sprintf("URL-based secret source from %s", us.config.URLTemplate)
+}
+
+func (us *urlSecrets) Get(key string) (string, error) {
+	ctx := context.Background()
+
+	if err := us.limiter.Wait(ctx); err != nil {
+		return "", fmt.Errorf("rate limiter error: %w", err)
+	}
+
+	// Replace {key} placeholder in URL template
+	url := strings.ReplaceAll(us.config.URLTemplate, "{key}", key)
+
+	method := us.config.Method
+	if method == "" {
+		method = http.MethodGet
+	}
+
+	req, err := http.NewRequestWithContext(ctx, method, url, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %w", err)
+	}
+
+	// Add headers
+	for k, v := range us.config.Headers {
+		req.Header.Set(k, v)
+	}
+
+	response, err := us.httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to get secret: %w", err)
+	}
+	defer func() { _ = response.Body.Close() }()
+
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("status code %d: %w", response.StatusCode, errFailedToGetSecret)
+	}
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response: %w", err)
+	}
```

Comment:
Do we have a maximum size for secrets set? Depending on it, it could make sense to use something like the `io.LimitReader` here to ensure we don't read bodies greater than that, which could help with memory usage, and I also have the intuition would be a better security practice?